central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -3,18 +3,82 @@ package datastore
 import (
 	"context"
 
+	"github.com/heimdalr/dag"
 	"github.com/stackrox/rox/central/resourcecollection/datastore/index"
 	"github.com/stackrox/rox/central/resourcecollection/datastore/search"
-	"github.com/stackrox/rox/central/resourcecollection/datastore/store/postgres"
+	"github.com/stackrox/rox/central/resourcecollection/datastore/store"
 	v1 "github.com/stackrox/rox/generated/api/v1"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/logging"
 	pkgSearch "github.com/stackrox/rox/pkg/search"
 )
 
+var (
+	log = logging.LoggerForModule()
+)
+
 type datastoreImpl struct {
-	storage  postgres.Store
+	storage  store.Store
 	indexer  index.Indexer
 	searcher search.Searcher
+
+	graph *dag.DAG
+}
+
+func (ds *datastoreImpl) initGraph(ctx context.Context) error {
+	if ds.graph != nil {
+		return nil
+	}
+	ds.graph = dag.NewDAG()
+
+	// first walk adds vertices
+	err := ds.storage.Walk(ctx, func(obj *storage.ResourceCollection) error {
+		return ds.graph.AddVertexByID(obj.GetId(), obj.GetId())
+	})
+	if err != nil {
+		return err
+	}
+
+	// second walk adds edges
+	return ds.storage.Walk(ctx, func(obj *storage.ResourceCollection) error {
```

Comment:
This feels inefficient because it's two walks through the data store. A simple solution would be to fetch it all and add it to an array (or use --get-all-func in gen.go to generate a GetAll func) and then traverse that to build the DAG. But there is a concern (however small of running out of memory).

So for the moment, do you mind creating a ticket and TODO to revisit optimizing this?

Reply:
This has been addressed in recent commits